core/processor/processor.go

```diff
@@ -0,0 +1,122 @@
+package processor
+
+import (
+	"context"
+	"sync"
+
+	"chain/database/pg"
+	"chain/log"
+	"chain/protocol"
+	"chain/protocol/bc"
+)
+
+type CursorStore struct {
+	DB pg.DB
+
+	mu      sync.Mutex
+	cursors map[string]*Cursor
+}
+
+func (ct *CursorStore) Cursor(name string) *Cursor {
+	ct.mu.Lock()
+	defer ct.mu.Unlock()
+	if ct.cursors == nil {
+		ct.cursors = make(map[string]*Cursor)
+	}
+	cursor, ok := ct.cursors[name]
+	if !ok {
+		cursor = newCursor(ct.DB, name, 0)
+		ct.cursors[name] = cursor
+	}
+	return cursor
+}
+
+func (ct *CursorStore) ProcessBlocks(ctx context.Context, c *protocol.Chain, cursorName string, cb func(context.Context, *bc.Block) error) {
+	cursor := ct.Cursor(cursorName)
+	for {
+		height := cursor.Height()
+		select {
+		case <-c.WaitForBlock(height + 1):
+			block, err := c.GetBlock(ctx, height+1)
+			if err != nil {
+				log.Error(ctx, err)
+				continue
+			}
+			err = cb(ctx, block)
+			if err != nil {
+				log.Error(ctx, err)
+				continue
+			}
+			// This could cause issues, since it is not inside of a
+			// database transaction.
```

Comment:
I think we should try to write the update with a where clause that ensures the value never decreases. We're already assuming that the callback is idempotent. Would that solve all the issues?